src/main/java/teammates/common/util/Version.java

```diff
@@ -0,0 +1,167 @@
+    }
+    
+    /**
+     * Converts Version to String in format XX.XX.XXXX
+     */
+    public String toString() {
+        String version = "";
+        if (major != null) {
+            version += String.valueOf(major);
+        }
+        if (minor != null) {
+            String minorInString = String.valueOf(minor);
+            if (minorInString.length() < 2) {
+                minorInString = "0" + minorInString;
+            }
+            version += "." + minorInString;
+        }
+        if (patch != null) {
+            version += "." + patch;
+        }
+        return version;
+    }
+    
+    /**
+     * Converts to String in format XX-XX-XXXX
+     */
+    public String toStringForQuery() {
+        return toString().replace('.', '-');
+    }
+    
+    /**
+     * Gets all available versions.
+     */
+    public static List<Version> getAvailableVersions() {
+        List<String> versionListInString = new ArrayList<String>(modulesService.getVersions(null)); // null == default module
+        List<Version> versionList = new ArrayList<Version>();
+        for(String versionInString : versionListInString) {
+            versionList.add(new Version(versionInString));
+        }
+        Collections.sort(versionList, new VersionComparator());
+        return versionList;
+    }
+    
+    /**
+     * Comparator for version strings.
+     * It sorts versions in DESCENDING order by major version then by minor version then by patch version
+     */
+    private static class VersionComparator implements Comparator<Version> {
+        @Override
+        public int compare(Version v1, Version v2) {
+            Integer major1 = v1.getMajorVersion();
+            Integer major2 = v2.getMajorVersion();
+            if (major1 == null && major2 == null) {
+                return 0;
+            }
+            if (major1 == null) {
+                return 1;
+            }
+            if (major2 == null) {
+                return -1;
+            }
+            if (!major1.equals(major2)) {
+                return -major1.compareTo(major2);
+            }
+            
+            Integer minor1 = v1.getMinorVersion();
+            Integer minor2 = v2.getMinorVersion();
+            if (minor1 == null && minor2 == null) {
+                return 0;
+            }
+            if (minor1 == null) {
+                return 1;
+            }
+            if (minor2 == null) {
+                return -1;
+            }
+            if (!minor1.equals(minor2)) {
+                return -minor1.compareTo(minor2);
+            }
+            
+            String patch1 = v1.getPatchVersion();
+            String patch2 = v2.getPatchVersion();
+            if (patch1 == null && patch2 == null) {
+                return 0;
+            }
+            if (patch1 == null) {
+                return 1;
+            }
+            if (patch2 == null) {
+                return -1;
+            }
+            return -patch1.compareTo(patch2);
```

Comment:
Looks very repetitive. Can the component-wise comparison be abstracted out into a common function?